File: zig/src/Token.zig
const Token = @This();

type: Type,
literal: []const u8,

pub const Type = enum {
    illegal,
    eof,

    ident,
    int,

    assign,
    plus,
    minus,
    bang,
    asterisk,
    slash,

    lt,
    gt,
    eq,
    ne,

    comma,
    semicolon,

    lparen,
    rparen,
    lbrace,
    rbrace,

    function,
    let,
    true,
    false,
    @"if",
    @"else",
    @"return",
};

pub const eof = Token{ .type = .eof, .literal = "" };
pub const assign = Token{ .type = .assign, .literal = "=" };
pub const plus = Token{ .type = .plus, .literal = "+" };
pub const minus = Token{ .type = .minus, .literal = "-" };
pub const bang = Token{ .type = .bang, .literal = "!" };
pub const asterisk = Token{ .type = .asterisk, .literal = "*" };
pub const slash = Token{ .type = .slash, .literal = "/" };
pub const lt = Token{ .type = .lt, .literal = "<" };
pub const gt = Token{ .type = .gt, .literal = ">" };
pub const eq = Token{ .type = .eq, .literal = "==" };
pub const ne = Token{ .type = .ne, .literal = "!=" };
pub const comma = Token{ .type = .comma, .literal = "," };
pub const semicolon = Token{ .type = .semicolon, .literal = ";" };
pub const lparen = Token{ .type = .lparen, .literal = "(" };
pub const rparen = Token{ .type = .rparen, .literal = ")" };
pub const lbrace = Token{ .type = .lbrace, .literal = "{" };
pub const rbrace = Token{ .type = .rbrace, .literal = "}" };
pub const function = Token{ .type = .function, .literal = "fn" };
pub const let = Token{ .type = .let, .literal = "let" };
pub const @"true" = Token{ .type = .true, .literal = "true" };
pub const @"false" = Token{ .type = .false, .literal = "false" };
pub const @"if" = Token{ .type = .@"if", .literal = "if" };
pub const @"else" = Token{ .type = .@"else", .literal = "else" };
pub const @"return" = Token{ .type = .@"return", .literal = "return" };

pub fn ident(l: []const u8) Token {
    return Token{ .type = .ident, .literal = l };
}

pub fn int(l: []const u8) Token {
    return Token{ .type = .int, .literal = l };
}

File: c/src/token.c
#pragma once

#include "string.c"

typedef enum TokenType {
  TOKEN_ILLEGAL,
  TOKEN_EOF,

  // identifiers and literals
  TOKEN_IDENT,
  TOKEN_INT,

  // operators
  TOKEN_ASSIGN,
  TOKEN_PLUS,
  TOKEN_MINUS,
  TOKEN_BANG,
  TOKEN_ASTERISK,
  TOKEN_SLASH,

  TOKEN_LT,
  TOKEN_GT,

  TOKEN_EQ,
  TOKEN_NOT_EQ,

  // delimiters
  TOKEN_COMMA,
  TOKEN_SEMICOLON,

  TOKEN_LPAREN,
  TOKEN_RPAREN,
  TOKEN_LBRACE,
  TOKEN_RBRACE,

  // keywords
  TOKEN_FUNCTION,
  TOKEN_LET,
  TOKEN_TRUE,
  TOKEN_FALSE,
  TOKEN_IF,
  TOKEN_ELSE,
  TOKEN_RETURN,
} TokenType;

const String token_type_strings[] = {
    String("ILLEGAL"),  String("EOF"),    String("IDENT"),  String("INT"),
    String("ASSIGN"),   String("PLUS"),   String("MINUS"),  String("BANG"),
    String("ASTERISK"), String("SLASH"),  String("LT"),     String("GT"),
    String("EQ"),       String("NOT_EQ"), String("COMMA"),  String("SEMICOLON"),
    String("LPAREN"),   String("RPAREN"), String("LBRACE"), String("RBRACE"),
    String("FUNCTION"), String("LET"),    String("TRUE"),   String("FALSE"),
    String("IF"),       String("ELSE"),   String("RETURN"),
};

typedef struct Token {
  TokenType type;
  String literal;
} Token;

TokenType token_type_from_ident(String ident) {
  if (string_cmp(ident, String("fn"))) {
    return TOKEN_FUNCTION;
  } else if (string_cmp(ident, String("let"))) {
    return TOKEN_LET;
  } else if (string_cmp(ident, String("true"))) {
    return TOKEN_TRUE;
  } else if (string_cmp(ident, String("false"))) {
    return TOKEN_FALSE;
  } else if (string_cmp(ident, String("if"))) {
    return TOKEN_IF;
  } else if (string_cmp(ident, String("else"))) {
    return TOKEN_ELSE;
  } else if (string_cmp(ident, String("return"))) {
    return TOKEN_RETURN;
  }
  return TOKEN_IDENT;
}
